core/src/main/scala/akka/kafka/scaladsl/Transactional.scala

```diff
@@ -51,24 +67,29 @@ object Transactional {
   ): Flow[Envelope[K, V, ConsumerMessage.PartitionOffset], Results[K, V, ConsumerMessage.PartitionOffset], NotUsed] = {
     require(transactionalId != null && transactionalId.length > 0, "You must define a Transactional id.")
 
-    val txSettings = settings.withProperties(
-      ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG -> true.toString,
-      ProducerConfig.TRANSACTIONAL_ID_CONFIG -> transactionalId,
-      ProducerConfig.MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION -> 1.toString
-    )
+    val txSettingsToProducer = (txid: String) => {
```

Comment:
So this creates a producer (with a unique `transactionalId`) for each partition that is assigned to this node. Isn't that expensive, or is the Kafka client smart about pooling the resources for those multiple producers?

Reply:
@raboof The original idea was to create a producer per partition, but you're right that it's possible that that could have unintended performance problems for large partition subscriptions.  AFAIK the separate instances of `KafkaProducer` don't share a common resource pool, so they would be managed separately.  We don't appear to have a precedent in our docs for using producers per substream with other partitioned sources.

I looked at Kafka Streams impl and the original design doc (https://docs.google.com/document/d/1pGZ8xtOOyGwDYgH5vA6h19zOMMaduFK1DAB8_gBYA2c/edit#heading=h.mki3gltx1zw) and they create 1 `KafkaProducer` per `Task`.  Each `TaskId` (1:1 with a `Task`) appears to represent one assigned partition.  The Kafka Streams `Task` `transactional.id` for a producer is set with a pattern of the application id (defined by the user across all instances), and a `TaskId`.  So my initial understanding is that it will indeed create one producer per partition.  I think we need to study this implementation more closely to understand if we need to be smarter about the number of producers we spawn.

Reply:
I took a look at this again and it does look like a producer is created per task.  I'm less clear on whether a task is 1:1 with a partition, but since you can only define one partition per task I don't see how it could be anything else.

This understanding conflicts with the original design doc, so I posed the question to the kafka-users mailing list.  Hopefully we'll get an authoritative reply.

https://lists.apache.org/thread.html/491b89840776498009fa773a737bf512a2c144558cce961b4ff8c768@%3Cusers.kafka.apache.org%3E